checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -353,9 +576,319 @@ protected boolean skipContractCheck(Tree tree, FlowExpressions.Receiver expr, Fl
+        }
+
+        // Perform the validity check and issue an error if not valid.
+
+        for (int i = 0; i < guardSatisfiedIndex.length; i++) {
+            if (guardSatisfiedIndex[i] != -1) {
+                for (int j = i + 1; j < guardSatisfiedIndex.length; j++) {
+                    if (guardSatisfiedIndex[i] == guardSatisfiedIndex[j]) {
+                        // The @GuardedBy annotations must be identical on the corresponding actual parameters.
+                        AnnotationMirror anno1 = passedArgAnnotations.get(i);
+                        AnnotationMirror anno2 = passedArgAnnotations.get(j);
+                        if (anno1 != null && anno2 != null) {
+                            boolean bothAreGSwithNoIndex = false;
+
+                                if (AnnotationUtils.areSameByClass(anno1, checkerGuardSatisfiedClass) &&
+                                    AnnotationUtils.areSameByClass(anno2, checkerGuardSatisfiedClass)){
+                                    if (AnnotationUtils.
+                                        getElementValue(anno1, "value", Integer.class, true) == -1 &&
+                                        AnnotationUtils.
+                                        getElementValue(anno2, "value", Integer.class, true) == -1) {
+                                        // Generally speaking, two @GuardSatisfied annotations with no index are incomparable.
+                                        // TODO: If they come from the same variable, they are comparable.  Fix and add a test case.
+                                        bothAreGSwithNoIndex = true;
+                                    }
+                                }
+
+                                if (bothAreGSwithNoIndex ||
+                                  !(atypeFactory.getQualifierHierarchy().isSubtype(anno1, anno2) ||
+                                  atypeFactory.getQualifierHierarchy().isSubtype(anno2, anno1))) {
+                                // TODO: allow these strings to be localized
+
+                                String formalParam1 = null;
+
+                                if (i == 0) {
+                                    formalParam1 = "The receiver type";
+                                } else {
+                                    formalParam1 = "Parameter #" + i; // i, not i-1, so the index is 1-based
+                                }
+
+                                String formalParam2 = "parameter #" + j; // j, not j-1, so the index is 1-based
+
+                                checker.report(Result.failure(
+                                        "guardsatisfied.parameters.must.match",
+                                        formalParam1, formalParam2, invokedMethod.toString(), guardSatisfiedIndex[i], anno1, anno2), node);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        return super.visitMethodInvocation(node, p);
```

Comment:
If I don't call super, then I will need to check before calling checker.report for each warning whether that warning is suppressed. Otherwise the user may end up seeing 0 warnings on code that is not sound. I think the safer approach is to call super - otherwise, any changes to the warning suppression mechanism would need to be made here as well.